### DISCUSSION TITLE
Axis rotation

### DISCUSSION THREAD

**Discussion by Kieruss** [StjerneIdioten/I3D-Blender-Addon#128] [Ideas & Feature Requests]
Problem: When you export from Blender and import to GE, the origin's X axis is rotated 90 degrees (essentially). Since so many things are axis dependent, this can be annoying and confusing. You first have to set up the axis in Blender with how you want it to appear in-game, and then, with "Affect only Origins" enabled, do R X -90 so that it looks wrong in Blender, but turns up in GE how you want it.

Solution: Option to rotate the axis/origin on export, so that the axis looks the same in Blender and GE

**Comment by StjerneIdioten** [comment]
I have a fairly long discussion over on the VertexDezign Discord about this exact topic. You do seem to already have some understanding of why, but I will give the whole explanation just for clarity. As well as tell you that what you seek is actually already implemented 😉  

### Coordinate Systems: Blender vs Giants Engine
Game engines and 3D-software all have some sort of base coordinate system defined which defines what direction is up/down, left/right and forward/backwards. But they don't necessarily agree on which of the axes **X, Y & Z** is what. To tell what kind of system a program uses it is normally defined as what axis is **Up**, combined with the **handedness** of the coordinate system (I believe the word stems from math/physics, where you can use your left or right hand to find out what kind of coordinate system you are dealing with) The following chart shows a few programs and the definition of their coordinate system.

![image](https://user-images.githubusercontent.com/11218511/112073873-3a458080-8b75-11eb-8816-73a6e624b4d3.png)

It can be seen that Blender falls into the camp of having a (Z-Up, right handed) coordinate system, which means that our models look like the following, when seen from a "Front Right" perspective (At least this is what I believe Blender categorises it as)

![image](https://user-images.githubusercontent.com/11218511/112075642-2bf96380-8b79-11eb-89be-ddc68ccb5e74.png)

Which of course makes sense, since I just modelled coloured arrows that correspond to their respective axis in Blender. Now GE on the other hand, falls into the same category as Maya in that previous chart. So it's world is defined as having (Y-Up, right handed), which is slightly different from Blenders. As you have noted it more or less corresponds to a 90° rotation around the X-axis (Due to sharing the X-axis by both being right handed)

If I maintain the same camera view, so we look at the world from the same perspective regardless of whether we are in Blender or GE. And I export the arrows with exactly the same coordinates as they have in Blender, then it looks like this in GE:

![image](https://user-images.githubusercontent.com/11218511/112076297-78916e80-8b7a-11eb-901b-ff92c13e2184.png)

If we were to look at the arrows transforms, then they would have exactly the same coordinates in Blender and GE at this point. But since GE is essentially rotated 90° around X, we can see that from the cameras perspective the arrows have been rotated as well. If we stick with the convention that the X-axis represents **Right** and we know that **Y** is **Up**, then the blue arrow which was pointing **Up** in Blender is now pointing  in the direction of **Backward**

### What the exporter does

To get rid of this inconsistency, the exporter transforms your objects coordinate basis in a way so that if something was pointing **Up** in Blender, it will also be pointing **Up** when it gets into GE. That will look like this:

![image](https://user-images.githubusercontent.com/11218511/112077368-a2e42b80-8b7c-11eb-9c67-9ae4ce5c41d8.png)

As you can see, from the perspective of the camera (which has the same view as the corresponding camera in Blender), the arrows are pointing in the exact same direction, but if we were to inspect the actual positions of the individual vertices it could be seen that they have changed their position.

### So why does the exporter do this?

No matter what you do, Blender and GE does not agree on the coordinate system. So at some point you are forced to make a choice. Do you want to keep the coordinates of the vertices 1:1 or do you want to keep the "view" 1:1 between the two programs. When I started work on this, I decided to keep the view 1:1. So **Up** in Blender is going to be your **Up** when it gets into GE. The reason for doing this is that Blender isn't really geared for handling another coordinate system. In Maya e.g. you can change your coordinate base and model 1:1 between Maya and GE. But in Blender it is hardcoded and thus all the views, UI grid and so on is geared for this Z-Up world. And as I imagined spending most of my time in Blender, this is how I chose to design it.

### The downside

Since this process changes the coordinate basis of your mesh and objects, this means that if you have positioned something to specifically follow a certain axis, then this will also be transformed. In the case of e.g. a lot of the empties (transformgroups) that you need for vehicles, they are used for things that rely on the orientation of the axis of the empty. This WILL break upon export, if great care isn't taken to consider their conversion, when you place them in Blender. 

The following table might be of some help:

World View | Blender  | Giants Engine
:------------:|:---------:|:--------------:
Up | Z | Y
Down | -Z | -Y
Left | -X | -X
Right | X | X 
Forward | Y | -Z
Backward | -Y | Z

So for example if you want your empty to point it's Z-axis in the direction of Z, when it gets into GE. You should point the Z-axis in the direction of the **Global Y** in Blender.

### What if I don't like this conversion?

You are in luck 😉 I thought of this when I originally implemented the conversion and I actually give you the option of converting your i3d into ANY coordinate system you desire.

![image](https://user-images.githubusercontent.com/11218511/112079892-3fa8c800-8b81-11eb-9781-daf69bdb5425.png)

These two options in the exporter panel gives you full control over the coordinate system transformation. They default to *(-Z Forward, Y-up)* which is the conversion that gives you this 1:1 view between Blender and GE. If you don't want this, you can change them to *(Y Forward, Z Up)* which tells the exporter to keep the coordinate system as it is. This does maintain your axis orientations, but beware that it makes things a bit more annoying to work with in Blender 😉 

### Future consideration

As a little ending remark I can share the thoughts that I have had about this and the solution I like the most, which is to do a sort of "best of both worlds" approach, where you get the option of toggling this on a per object basis. So that you could have your models following the "view", since axis orientation mostly doesn't matter for visuals. And then you could turn of the conversion for your empties, so that you can position them correctly directly in Blender, without having to keep tracking of axes conversions.

I am open to any other suggestions though, if someone has a better way to do things.

**Reply by Kieruss** [reply]
I did play with those Forward/Up settings, but they cause the model to rotate as well. What I'm suggesting is to have the exporter perform an origin-only rotation of -90 degrees on the local X axis. Doing this with the standard -Z Forward, Y UP export settings results in the model being oriented the same way AND the axis being oriented the same as in Blender.

I start with this in Blender
![image](https://user-images.githubusercontent.com/24835155/112090411-9943dd80-8be7-11eb-8c6b-58c1871af8d7.png)

and get this in GE (I did the origin -90 on local X on each shape after taking the first screenshot)
![image](https://user-images.githubusercontent.com/24835155/112090462-b4165200-8be7-11eb-8eef-670306130391.png)
![image](https://user-images.githubusercontent.com/24835155/112090488-c2fd0480-8be7-11eb-875e-b2e3c93be7b3.png)
![image](https://user-images.githubusercontent.com/24835155/112090512-cdb79980-8be7-11eb-8dad-af8eddfe6f8c.png)

**Reply by Kieruss** [reply]
Here it is again, expressed in arrows :)
![image](https://user-images.githubusercontent.com/24835155/112093690-078b9e80-8bee-11eb-9057-111867202a9a.png)
![image](https://user-images.githubusercontent.com/24835155/112093957-6d782600-8bee-11eb-9b77-27500a975a87.png)

If I use the Z Up, -X Forward (without doing my local rotation thing), I end up with this.. The shape X is up, and the axis Y is still up.. So, to my mind, it's wrong in every important aspect :)
![image](https://user-images.githubusercontent.com/24835155/112094331-30f8fa00-8bef-11eb-8143-0ebc260a4567.png)

**Reply by StjerneIdioten** [reply]
> If I use the Z Up, -X Forward (without doing my local rotation thing), I end up with this.. The shape X is up, and the axis Y is still up.. So, to my mind, it's wrong in every important aspect :)

I do agree on that, but that is only because you used -X as forward instead of Y 😛 Using the proper input gives you this:

![image](https://user-images.githubusercontent.com/11218511/112176218-34dd4a00-8bf8-11eb-92ec-003f743822ce.png)

Which we can agree on being "wrong" in the sense of the "view" of your model, but correct in the sense of being aligned with the original axes, so only half-way wrong 🤣 

For the transformation you show I think I get it. You do -90° on X of the origins and get this:

![image](https://user-images.githubusercontent.com/11218511/112179172-bdf58080-8bfa-11eb-8e9d-40b6952c426e.png)

Which has a -90 Rotate X inside of GE, but everything now lines up.

Now try and do that with something like this:

![image](https://user-images.githubusercontent.com/11218511/112181540-d36baa00-8bfc-11eb-93c4-33e780480e64.png)

I have parented the arrows axis-wise. Try and rotate origin now and be in for a world of pain 🤣 It is not impossible to solve of course, especially not programmatically. It is just an extra thing to take into consideration, that if you have child objects, but still want them positioned in relation to the mesh as they were. Then when you rotate the origin, you have to convert the children as well since e.g. the child of the Y-arrow now has to be offset by 6m on the Z-axis instead of the Y-axis. To still be in the same position.

This isn't made any easier by Blenders whole "inverse parent transform" concept of an invisible extra transform that relates things between parent and child.

It is definitely something I can try and have another look at, when I rewrite that part of the code.

**Reply by Kieruss** [reply]
I'm a bit confused about how you do it currently, by default, without affecting the children. Exporting your pic would have the arrows looking the same (blue points away from ground), but the Y axis would be up, rather than Z.... and the children would be parented on a different axis (the child blue is now a child of parent blue's Y)... How can you do this, but having Z up causes the shapes to rotate?

**Comment by Kieruss** [comment]
![image](https://user-images.githubusercontent.com/24835155/148261376-acdc4473-cf5a-4b12-806f-cefca2f90d58.png)
It took me a while, but this checkbox lets me rotate origins of parent or child, without upsetting either. Not sure if that helps programmatic rotation :)
